{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ETL for Ingesting Data from JSON Sources\n",
    "\n",
    "This notebook outlines how you can transform NGS data to suit ingestion to Beacon in the data portal.\n",
    "\n",
    "We can start by import basic imports required for this data transformation.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataset Entry\n",
    "\n",
    "A dataset entry is essential to encapsulate information regarding the collection of information. Let's create one using some made up date. Please note that relevant field should be populated in practice using the actual data you harvest during the data collection procedures.\n",
    "\n",
    "We should also have information regarding the reference assembly used to call the variants (eg.: GRCH37/38)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = {\n",
    "    \"createDateTime\": \"2021-03-21T02:37:00-08:00\",\n",
    "    \"dataUseConditions\": {\n",
    "        \"duoDataUse\": [\n",
    "            {\n",
    "                \"id\": \"DUO:0000042\",\n",
    "                \"label\": \"general research use\",\n",
    "                \"version\": \"17-07-2016\"\n",
    "            }\n",
    "        ]\n",
    "    },\n",
    "    \"description\": \"Simulation set 1.\",\n",
    "    \"externalUrl\": \"http://example.org/wiki/Main_Page\",\n",
    "    \"info\": {},\n",
    "    \"name\": \"Dataset with NGS data\",\n",
    "    \"updateDateTime\": \"2022-08-05T17:21:00+01:00\",\n",
    "    \"version\": \"v1.1\"\n",
    "}\n",
    "assembly_id = \"GRCH38\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Individuals Information\n",
    "\n",
    "Each VCF sample must have an origin individual. These entries can be constructed as follows. Note that for this example, we are using four individual entries.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "individuals = [\n",
    "    {\n",
    "      \"id\": \"INDIVIDUAL_1\",\n",
    "      \"karyotypicSex\": \"XX\",\n",
    "      \"sex\": {\n",
    "        \"id\": \"SNOMED:248153007\",\n",
    "        \"label\": \"MALE\"\n",
    "      }\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"INDIVIDUAL_2\",\n",
    "      \"karyotypicSex\": \"XY\",\n",
    "      \"sex\": {\n",
    "        \"id\": \"SNOMED:248152002\",\n",
    "        \"label\": \"FEMALE\"\n",
    "      }\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"INDIVIDUAL_3\",\n",
    "      \"karyotypicSex\": \"XX\",\n",
    "      \"sex\": {\n",
    "        \"id\": \"SNOMED:248153007\",\n",
    "        \"label\": \"MALE\"\n",
    "      }\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"INDIVIDUAL_4\",\n",
    "      \"karyotypicSex\": \"XY\",\n",
    "      \"sex\": {\n",
    "        \"id\": \"SNOMED:248152002\",\n",
    "        \"label\": \"FEMALE\"\n",
    "      }\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Biosample Information\n",
    "\n",
    "Biosample encapsulates information regarding the collected physiological sample from the individuals body. For each individual, this example related exactly one biosamples.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "biosamples = [\n",
    "    {\n",
    "      \"id\": \"BIOSAMPLE_1\",\n",
    "      \"individualId\": \"INDIVIDUAL_1\",\n",
    "      \"biosampleStatus\": {\n",
    "        \"id\": \"SNOMED:365641003\",\n",
    "        \"label\": \"Minor blood groups - finding\"\n",
    "      },\n",
    "      \"collectionDate\": \"2020-1-1\",\n",
    "      \"sampleOriginType\": {\n",
    "        \"id\": \"SNOMED:31675002\",\n",
    "        \"label\": \"Capillary blood\"\n",
    "      },\n",
    "      \"sampleOriginDetail\": {\n",
    "        \"id\": \"SNOMED:31675002\",\n",
    "        \"label\": \"Capillary blood\"\n",
    "      },\n",
    "      \"info\": {},\n",
    "      \"notes\": \"\"\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"BIOSAMPLE_2\",\n",
    "      \"individualId\": \"INDIVIDUAL_2\",\n",
    "      \"biosampleStatus\": {\n",
    "        \"id\": \"SNOMED:365641003\",\n",
    "        \"label\": \"Minor blood groups - finding\"\n",
    "      },\n",
    "      \"collectionDate\": \"2020-1-2\",\n",
    "      \"sampleOriginType\": {\n",
    "        \"id\": \"SNOMED:31675002\",\n",
    "        \"label\": \"Capillary blood\"\n",
    "      },\n",
    "      \"sampleOriginDetail\": {\n",
    "        \"id\": \"SNOMED:31675002\",\n",
    "        \"label\": \"Capillary blood\"\n",
    "      },\n",
    "      \"info\": {},\n",
    "      \"notes\": \"\"\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"BIOSAMPLE_3\",\n",
    "      \"individualId\": \"INDIVIDUAL_3\",\n",
    "      \"biosampleStatus\": {\n",
    "        \"id\": \"SNOMED:365641003\",\n",
    "        \"label\": \"Minor blood groups - finding\"\n",
    "      },\n",
    "      \"collectionDate\": \"2020-1-3\",\n",
    "      \"sampleOriginType\": {\n",
    "        \"id\": \"SNOMED:31675002\",\n",
    "        \"label\": \"Capillary blood\"\n",
    "      },\n",
    "      \"sampleOriginDetail\": {\n",
    "        \"id\": \"SNOMED:31675002\",\n",
    "        \"label\": \"Capillary blood\"\n",
    "      },\n",
    "      \"info\": {},\n",
    "      \"notes\": \"\"\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"BIOSAMPLE_4\",\n",
    "      \"individualId\": \"INDIVIDUAL_4\",\n",
    "      \"biosampleStatus\": {\n",
    "        \"id\": \"SNOMED:365641003\",\n",
    "        \"label\": \"Minor blood groups - finding\"\n",
    "      },\n",
    "      \"collectionDate\": \"2020-1-4\",\n",
    "      \"sampleOriginType\": {\n",
    "        \"id\": \"SNOMED:31675002\",\n",
    "        \"label\": \"Capillary blood\"\n",
    "      },\n",
    "      \"sampleOriginDetail\": {\n",
    "        \"id\": \"SNOMED:31675002\",\n",
    "        \"label\": \"Capillary blood\"\n",
    "      },\n",
    "      \"info\": {},\n",
    "      \"notes\": \"\"\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Information\n",
    "\n",
    "Run encapsulates information regarding sequencing run associated with the biosample.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "runs = [\n",
    "    {\n",
    "      \"id\": \"RUN_1\",\n",
    "      \"biosampleId\": \"BIOSAMPLE_1\",\n",
    "      \"individualId\": \"INDIVIDUAL_1\",\n",
    "      \"libraryLayout\": \"PAIRED\",\n",
    "      \"librarySelection\": \"RANDOM\",  \n",
    "      \"libraryStrategy\": \"WGS\",\n",
    "      \"platform\": \"Illumina\", \n",
    "      \"platformModel\": {\n",
    "        \"id\": \"OBI:0002630\",\n",
    "        \"label\": \"Illumina NovaSeq 6000\"\n",
    "      },\n",
    "      \"librarySource\": {\n",
    "        \"id\": \"GENEPIO:0001969\",\n",
    "        \"label\": \"other library source\"\n",
    "      },\n",
    "      \"runDate\": \"2019-1-1\"\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"RUN_2\",\n",
    "      \"biosampleId\": \"BIOSAMPLE_2\",\n",
    "      \"individualId\": \"INDIVIDUAL_2\",\n",
    "      \"libraryLayout\": \"PAIRED\",\n",
    "      \"librarySelection\": \"RANDOM\",  \n",
    "      \"libraryStrategy\": \"WGS\",\n",
    "      \"platform\": \"Illumina\", \n",
    "      \"platformModel\": {\n",
    "        \"id\": \"OBI:0002630\",\n",
    "        \"label\": \"Illumina NovaSeq 6000\"\n",
    "      },\n",
    "      \"librarySource\": {\n",
    "        \"id\": \"GENEPIO:0001969\",\n",
    "        \"label\": \"other library source\"\n",
    "      },\n",
    "      \"runDate\": \"2019-1-2\"\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"RUN_3\",\n",
    "      \"biosampleId\": \"BIOSAMPLE_3\",\n",
    "      \"individualId\": \"INDIVIDUAL_3\",\n",
    "      \"libraryLayout\": \"PAIRED\",\n",
    "      \"librarySelection\": \"RANDOM\",  \n",
    "      \"libraryStrategy\": \"WGS\",\n",
    "      \"platform\": \"Illumina\", \n",
    "      \"platformModel\": {\n",
    "        \"id\": \"OBI:0002630\",\n",
    "        \"label\": \"Illumina NovaSeq 6000\"\n",
    "      },\n",
    "      \"librarySource\": {\n",
    "        \"id\": \"GENEPIO:0001969\",\n",
    "        \"label\": \"other library source\"\n",
    "      },\n",
    "      \"runDate\": \"2019-1-3\"\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"RUN_4\",\n",
    "      \"biosampleId\": \"BIOSAMPLE_4\",\n",
    "      \"individualId\": \"INDIVIDUAL_4\",\n",
    "      \"libraryLayout\": \"PAIRED\",\n",
    "      \"librarySelection\": \"RANDOM\",  \n",
    "      \"libraryStrategy\": \"WGS\",\n",
    "      \"platform\": \"Illumina\", \n",
    "      \"platformModel\": {\n",
    "        \"id\": \"OBI:0002630\",\n",
    "        \"label\": \"Illumina NovaSeq 6000\"\n",
    "      },\n",
    "      \"librarySource\": {\n",
    "        \"id\": \"GENEPIO:0001969\",\n",
    "        \"label\": \"other library source\"\n",
    "      },\n",
    "      \"runDate\": \"2019-1-4\"\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis Information\n",
    "\n",
    "Information regarding the bioinformatics analysis comes under this entity information.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyses = [\n",
    "    {\n",
    "      \"id\": \"ANALYSIS_1\",\n",
    "      \"individualId\": \"INDIVIDUAL_1\",\n",
    "      \"biosampleId\": \"BIOSAMPLE_1\",\n",
    "      \"runId\": \"RUN_1\",\n",
    "      \"analysisDate\": \"2019-1-1\",\n",
    "      \"pipelineName\": \"GATK Best Practices\",\n",
    "      \"pipelineRef\": \"Broad Institute\",\n",
    "      \"variantCaller\": \"GATK HaplotypeCaller\",\n",
    "      \"vcfSampleId\": \"SMPL_1\"\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"ANALYSIS_2\",\n",
    "      \"individualId\": \"INDIVIDUAL_2\",\n",
    "      \"biosampleId\": \"BIOSAMPLE_2\",\n",
    "      \"runId\": \"RUN_2\",\n",
    "      \"analysisDate\": \"2019-1-2\",\n",
    "      \"pipelineName\": \"GATK Best Practices\",\n",
    "      \"pipelineRef\": \"Broad Institute\",\n",
    "      \"variantCaller\": \"GATK HaplotypeCaller\",\n",
    "      \"vcfSampleId\": \"SMPL_2\"\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"ANALYSIS_3\",\n",
    "      \"individualId\": \"INDIVIDUAL_3\",\n",
    "      \"biosampleId\": \"BIOSAMPLE_3\",\n",
    "      \"runId\": \"RUN_3\",\n",
    "      \"analysisDate\": \"2019-1-3\",\n",
    "      \"pipelineName\": \"GATK Best Practices\",\n",
    "      \"pipelineRef\": \"Broad Institute\",\n",
    "      \"variantCaller\": \"GATK HaplotypeCaller\",\n",
    "      \"vcfSampleId\": \"SMPL_3\"\n",
    "    },\n",
    "    {\n",
    "      \"id\": \"ANALYSIS_4\",\n",
    "      \"individualId\": \"INDIVIDUAL_4\",\n",
    "      \"biosampleId\": \"BIOSAMPLE_4\",\n",
    "      \"runId\": \"RUN_4\",\n",
    "      \"analysisDate\": \"2019-1-4\",\n",
    "      \"pipelineName\": \"GATK Best Practices\",\n",
    "      \"pipelineRef\": \"Broad Institute\",\n",
    "      \"variantCaller\": \"GATK HaplotypeCaller\",\n",
    "      \"vcfSampleId\": \"SMPL_4\"\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = \"\"\"\n",
    "##fileformat=VCFv4.2\n",
    "##contig=<ID=chr2,length=242193529>\n",
    "##contig=<ID=chr3,length=198295559>\n",
    "##INFO=<ID=NS,Number=1,Type=Integer,Description=\"Number of Samples With Data\">\n",
    "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Total Depth\">\n",
    "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
    "#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tSMPL_1\\tSMPL_2\\tSMPL_3\\tSMPL_4\n",
    "chr2\\t20436\\t.\\tC\\tT\\t47\\tPASS\\tNS=4;DP=18\\tGT\\t1|1\\t1|0\\t0|1\\t1|0\n",
    "chr2\\t42619\\t.\\tG\\tT\\t23\\tPASS\\tNS=4;DP=30\\tGT\\t0|1\\t1|0\\t0|0\\t0|1\n",
    "chr3\\t23398\\t.\\tG\\tA\\t60\\tPASS\\tNS=4;DP=70\\tGT\\t1|0\\t1|1\\t1|1\\t0|1\n",
    "chr3\\t30264\\t.\\tA\\tT\\t30\\tPASS\\tNS=4;DP=79\\tGT\\t0|0\\t0|1\\t0|1\\t1|1\n",
    "chr3\\t47708\\t.\\tT\\tG\\t25\\tPASS\\tNS=4;DP=31\\tGT\\t1|0\\t1|1\\t0|1\\t0|0\n",
    "\"\"\".strip()\n",
    "\n",
    "with open(\"data.vcf\", \"w\") as f:\n",
    "    f.write(vcf)\n",
    "\n",
    "!bgzip -f data.vcf\n",
    "!tabix -f data.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"dataset\": {\n",
      "    \"createDateTime\": \"2021-03-21T02:37:00-08:00\",\n",
      "    \"dataUseConditions\": {\n",
      "      \"duoDataUse\": [\n",
      "        {\n",
      "          \"id\": \"DUO:0000042\",\n",
      "          \"label\": \"general research use\",\n",
      "          \"version\": \"17-07-2016\"\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    \"description\": \"Simulation set 1.\",\n",
      "    \"externalUrl\": \"http://example.org/wiki/Main_Page\",\n",
      "    \"info\": {},\n",
      "    \"name\": \"Dataset with NGS data\",\n",
      "    \"updateDateTime\": \"2022-08-05T17:21:00+01:00\",\n",
      "    \"version\": \"v1.1\"\n",
      "  },\n",
      "  \"assemblyId\": \"GRCH38\",\n",
      "  \"individuals\": [\n",
      "    {\n",
      "      \"id\": \"INDIVIDUAL_1\",\n",
      "      \"karyotypicSex\": \"XX\",\n",
      "      \"sex\": {\n",
      "        \"id\": \"SNOMED:248153007\",\n",
      "        \"label\": \"MALE\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"INDIVIDUAL_2\",\n",
      "      \"karyotypicSex\": \"XY\",\n",
      "      \"sex\": {\n",
      "        \"id\": \"SNOMED:248152002\",\n",
      "        \"label\": \"FEMALE\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"INDIVIDUAL_3\",\n",
      "      \"karyotypicSex\": \"XX\",\n",
      "      \"sex\": {\n",
      "        \"id\": \"SNOMED:248153007\",\n",
      "        \"label\": \"MALE\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"INDIVIDUAL_4\",\n",
      "      \"karyotypicSex\": \"XY\",\n",
      "      \"sex\": {\n",
      "        \"id\": \"SNOMED:248152002\",\n",
      "        \"label\": \"FEMALE\"\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"biosamples\": [\n",
      "    {\n",
      "      \"id\": \"BIOSAMPLE_1\",\n",
      "      \"individualId\": \"INDIVIDUAL_1\",\n",
      "      \"biosampleStatus\": {\n",
      "        \"id\": \"SNOMED:365641003\",\n",
      "        \"label\": \"Minor blood groups - finding\"\n",
      "      },\n",
      "      \"collectionDate\": \"2020-1-1\",\n",
      "      \"sampleOriginType\": {\n",
      "        \"id\": \"SNOMED:31675002\",\n",
      "        \"label\": \"Capillary blood\"\n",
      "      },\n",
      "      \"sampleOriginDetail\": {\n",
      "        \"id\": \"SNOMED:31675002\",\n",
      "        \"label\": \"Capillary blood\"\n",
      "      },\n",
      "      \"info\": {},\n",
      "      \"notes\": \"\"\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"BIOSAMPLE_2\",\n",
      "      \"individualId\": \"INDIVIDUAL_2\",\n",
      "      \"biosampleStatus\": {\n",
      "        \"id\": \"SNOMED:365641003\",\n",
      "        \"label\": \"Minor blood groups - finding\"\n",
      "      },\n",
      "      \"collectionDate\": \"2020-1-2\",\n",
      "      \"sampleOriginType\": {\n",
      "        \"id\": \"SNOMED:31675002\",\n",
      "        \"label\": \"Capillary blood\"\n",
      "      },\n",
      "      \"sampleOriginDetail\": {\n",
      "        \"id\": \"SNOMED:31675002\",\n",
      "        \"label\": \"Capillary blood\"\n",
      "      },\n",
      "      \"info\": {},\n",
      "      \"notes\": \"\"\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"BIOSAMPLE_3\",\n",
      "      \"individualId\": \"INDIVIDUAL_3\",\n",
      "      \"biosampleStatus\": {\n",
      "        \"id\": \"SNOMED:365641003\",\n",
      "        \"label\": \"Minor blood groups - finding\"\n",
      "      },\n",
      "      \"collectionDate\": \"2020-1-3\",\n",
      "      \"sampleOriginType\": {\n",
      "        \"id\": \"SNOMED:31675002\",\n",
      "        \"label\": \"Capillary blood\"\n",
      "      },\n",
      "      \"sampleOriginDetail\": {\n",
      "        \"id\": \"SNOMED:31675002\",\n",
      "        \"label\": \"Capillary blood\"\n",
      "      },\n",
      "      \"info\": {},\n",
      "      \"notes\": \"\"\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"BIOSAMPLE_4\",\n",
      "      \"individualId\": \"INDIVIDUAL_4\",\n",
      "      \"biosampleStatus\": {\n",
      "        \"id\": \"SNOMED:365641003\",\n",
      "        \"label\": \"Minor blood groups - finding\"\n",
      "      },\n",
      "      \"collectionDate\": \"2020-1-4\",\n",
      "      \"sampleOriginType\": {\n",
      "        \"id\": \"SNOMED:31675002\",\n",
      "        \"label\": \"Capillary blood\"\n",
      "      },\n",
      "      \"sampleOriginDetail\": {\n",
      "        \"id\": \"SNOMED:31675002\",\n",
      "        \"label\": \"Capillary blood\"\n",
      "      },\n",
      "      \"info\": {},\n",
      "      \"notes\": \"\"\n",
      "    }\n",
      "  ],\n",
      "  \"runs\": [\n",
      "    {\n",
      "      \"id\": \"RUN_1\",\n",
      "      \"biosampleId\": \"BIOSAMPLE_1\",\n",
      "      \"individualId\": \"INDIVIDUAL_1\",\n",
      "      \"libraryLayout\": \"PAIRED\",\n",
      "      \"librarySelection\": \"RANDOM\",\n",
      "      \"libraryStrategy\": \"WGS\",\n",
      "      \"platform\": \"Illumina\",\n",
      "      \"platformModel\": {\n",
      "        \"id\": \"OBI:0002630\",\n",
      "        \"label\": \"Illumina NovaSeq 6000\"\n",
      "      },\n",
      "      \"librarySource\": {\n",
      "        \"id\": \"GENEPIO:0001969\",\n",
      "        \"label\": \"other library source\"\n",
      "      },\n",
      "      \"runDate\": \"2019-1-1\"\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"RUN_2\",\n",
      "      \"biosampleId\": \"BIOSAMPLE_2\",\n",
      "      \"individualId\": \"INDIVIDUAL_2\",\n",
      "      \"libraryLayout\": \"PAIRED\",\n",
      "      \"librarySelection\": \"RANDOM\",\n",
      "      \"libraryStrategy\": \"WGS\",\n",
      "      \"platform\": \"Illumina\",\n",
      "      \"platformModel\": {\n",
      "        \"id\": \"OBI:0002630\",\n",
      "        \"label\": \"Illumina NovaSeq 6000\"\n",
      "      },\n",
      "      \"librarySource\": {\n",
      "        \"id\": \"GENEPIO:0001969\",\n",
      "        \"label\": \"other library source\"\n",
      "      },\n",
      "      \"runDate\": \"2019-1-2\"\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"RUN_3\",\n",
      "      \"biosampleId\": \"BIOSAMPLE_3\",\n",
      "      \"individualId\": \"INDIVIDUAL_3\",\n",
      "      \"libraryLayout\": \"PAIRED\",\n",
      "      \"librarySelection\": \"RANDOM\",\n",
      "      \"libraryStrategy\": \"WGS\",\n",
      "      \"platform\": \"Illumina\",\n",
      "      \"platformModel\": {\n",
      "        \"id\": \"OBI:0002630\",\n",
      "        \"label\": \"Illumina NovaSeq 6000\"\n",
      "      },\n",
      "      \"librarySource\": {\n",
      "        \"id\": \"GENEPIO:0001969\",\n",
      "        \"label\": \"other library source\"\n",
      "      },\n",
      "      \"runDate\": \"2019-1-3\"\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"RUN_4\",\n",
      "      \"biosampleId\": \"BIOSAMPLE_4\",\n",
      "      \"individualId\": \"INDIVIDUAL_4\",\n",
      "      \"libraryLayout\": \"PAIRED\",\n",
      "      \"librarySelection\": \"RANDOM\",\n",
      "      \"libraryStrategy\": \"WGS\",\n",
      "      \"platform\": \"Illumina\",\n",
      "      \"platformModel\": {\n",
      "        \"id\": \"OBI:0002630\",\n",
      "        \"label\": \"Illumina NovaSeq 6000\"\n",
      "      },\n",
      "      \"librarySource\": {\n",
      "        \"id\": \"GENEPIO:0001969\",\n",
      "        \"label\": \"other library source\"\n",
      "      },\n",
      "      \"runDate\": \"2019-1-4\"\n",
      "    }\n",
      "  ],\n",
      "  \"analyses\": [\n",
      "    {\n",
      "      \"id\": \"ANALYSIS_1\",\n",
      "      \"individualId\": \"INDIVIDUAL_1\",\n",
      "      \"biosampleId\": \"BIOSAMPLE_1\",\n",
      "      \"runId\": \"RUN_1\",\n",
      "      \"analysisDate\": \"2019-1-1\",\n",
      "      \"pipelineName\": \"GATK Best Practices\",\n",
      "      \"pipelineRef\": \"Broad Institute\",\n",
      "      \"variantCaller\": \"GATK HaplotypeCaller\",\n",
      "      \"vcfSampleId\": \"SMPL_1\"\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"ANALYSIS_2\",\n",
      "      \"individualId\": \"INDIVIDUAL_2\",\n",
      "      \"biosampleId\": \"BIOSAMPLE_2\",\n",
      "      \"runId\": \"RUN_2\",\n",
      "      \"analysisDate\": \"2019-1-2\",\n",
      "      \"pipelineName\": \"GATK Best Practices\",\n",
      "      \"pipelineRef\": \"Broad Institute\",\n",
      "      \"variantCaller\": \"GATK HaplotypeCaller\",\n",
      "      \"vcfSampleId\": \"SMPL_2\"\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"ANALYSIS_3\",\n",
      "      \"individualId\": \"INDIVIDUAL_3\",\n",
      "      \"biosampleId\": \"BIOSAMPLE_3\",\n",
      "      \"runId\": \"RUN_3\",\n",
      "      \"analysisDate\": \"2019-1-3\",\n",
      "      \"pipelineName\": \"GATK Best Practices\",\n",
      "      \"pipelineRef\": \"Broad Institute\",\n",
      "      \"variantCaller\": \"GATK HaplotypeCaller\",\n",
      "      \"vcfSampleId\": \"SMPL_3\"\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"ANALYSIS_4\",\n",
      "      \"individualId\": \"INDIVIDUAL_4\",\n",
      "      \"biosampleId\": \"BIOSAMPLE_4\",\n",
      "      \"runId\": \"RUN_4\",\n",
      "      \"analysisDate\": \"2019-1-4\",\n",
      "      \"pipelineName\": \"GATK Best Practices\",\n",
      "      \"pipelineRef\": \"Broad Institute\",\n",
      "      \"variantCaller\": \"GATK HaplotypeCaller\",\n",
      "      \"vcfSampleId\": \"SMPL_4\"\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "submission = {\n",
    "    \"dataset\": dataset,\n",
    "    \"assemblyId\": assembly_id,\n",
    "    \"individuals\": individuals,\n",
    "    \"biosamples\": biosamples,\n",
    "    \"runs\": runs,\n",
    "    \"analyses\": analyses\n",
    "}\n",
    "\n",
    "print(json.dumps(submission, indent=2))\n",
    "json.dump(submission, open(\"ngs-submission.json\", \"w+\"), indent=2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validating the Submission\n",
    "\n",
    "Now you can validate the generated schema using the following scripts.\n",
    "\n",
    "Firstly we install python library `jsonschemas` which will help us validate the created payloads.\n",
    "\n",
    "We then download latest schemas to be used with `jsonschemas` library.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: jsonschema in /Users/wic053/miniforge3/envs/py312/lib/python3.12/site-packages (4.18.0)\n",
      "Requirement already satisfied: attrs>=22.2.0 in /Users/wic053/miniforge3/envs/py312/lib/python3.12/site-packages (from jsonschema) (23.2.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /Users/wic053/miniforge3/envs/py312/lib/python3.12/site-packages (from jsonschema) (2023.12.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in /Users/wic053/miniforge3/envs/py312/lib/python3.12/site-packages (from jsonschema) (0.35.1)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in /Users/wic053/miniforge3/envs/py312/lib/python3.12/site-packages (from jsonschema) (0.19.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install jsonschema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: schemas: File exists\n",
      "--2025-02-05 14:26:52--  https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/analysis-schema.json\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8002::154, 2606:50c0:8003::154, 2606:50c0:8000::154, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8002::154|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 3144 (3.1K) [text/plain]\n",
      "Saving to: ‘schemas/analysis-schema.json’\n",
      "\n",
      "schemas/analysis-sc 100%[===================>]   3.07K  --.-KB/s    in 0s      \n",
      "\n",
      "2025-02-05 14:26:52 (28.3 MB/s) - ‘schemas/analysis-schema.json’ saved [3144/3144]\n",
      "\n",
      "--2025-02-05 14:26:53--  https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/biosample-schema.json\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8003::154, 2606:50c0:8000::154, 2606:50c0:8001::154, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8003::154|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 120210 (117K) [text/plain]\n",
      "Saving to: ‘schemas/biosample-schema.json’\n",
      "\n",
      "schemas/biosample-s 100%[===================>] 117.39K  --.-KB/s    in 0.07s   \n",
      "\n",
      "2025-02-05 14:26:53 (1.56 MB/s) - ‘schemas/biosample-schema.json’ saved [120210/120210]\n",
      "\n",
      "--2025-02-05 14:26:53--  https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/dataset-schema.json\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8002::154, 2606:50c0:8003::154, 2606:50c0:8000::154, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8002::154|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 14639 (14K) [text/plain]\n",
      "Saving to: ‘schemas/dataset-schema.json’\n",
      "\n",
      "schemas/dataset-sch 100%[===================>]  14.30K  --.-KB/s    in 0.003s  \n",
      "\n",
      "2025-02-05 14:26:53 (4.46 MB/s) - ‘schemas/dataset-schema.json’ saved [14639/14639]\n",
      "\n",
      "--2025-02-05 14:26:54--  https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/individual-schema.json\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8002::154, 2606:50c0:8003::154, 2606:50c0:8000::154, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8002::154|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 184429 (180K) [text/plain]\n",
      "Saving to: ‘schemas/individual-schema.json’\n",
      "\n",
      "schemas/individual- 100%[===================>] 180.11K  --.-KB/s    in 0.1s    \n",
      "\n",
      "2025-02-05 14:26:54 (1.19 MB/s) - ‘schemas/individual-schema.json’ saved [184429/184429]\n",
      "\n",
      "--2025-02-05 14:26:54--  https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/run-schema.json\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8002::154, 2606:50c0:8003::154, 2606:50c0:8000::154, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8002::154|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 5308 (5.2K) [text/plain]\n",
      "Saving to: ‘schemas/run-schema.json’\n",
      "\n",
      "schemas/run-schema. 100%[===================>]   5.18K  --.-KB/s    in 0s      \n",
      "\n",
      "2025-02-05 14:26:55 (44.4 MB/s) - ‘schemas/run-schema.json’ saved [5308/5308]\n",
      "\n",
      "--2025-02-05 14:26:55--  https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/submit-dataset-schema-new.json\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8002::154, 2606:50c0:8003::154, 2606:50c0:8000::154, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8002::154|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2364 (2.3K) [text/plain]\n",
      "Saving to: ‘schemas/submit-dataset-schema-new.json’\n",
      "\n",
      "schemas/submit-data 100%[===================>]   2.31K  --.-KB/s    in 0s      \n",
      "\n",
      "2025-02-05 14:26:55 (25.3 MB/s) - ‘schemas/submit-dataset-schema-new.json’ saved [2364/2364]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%mkdir schemas\n",
    "!wget https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/analysis-schema.json -O schemas/analysis-schema.json    \n",
    "!wget https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/biosample-schema.json -O schemas/biosample-schema.json\n",
    "!wget https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/dataset-schema.json -O schemas/dataset-schema.json\n",
    "!wget https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/individual-schema.json -O schemas/individual-schema.json\n",
    "!wget https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/run-schema.json -O schemas/run-schema.json\n",
    "!wget https://raw.githubusercontent.com/GSI-Xapiens-CSIRO/sBeacon-BGSi/refs/heads/main/shared_resources/schemas/submit-dataset-schema-new.json -O schemas/submit-dataset-schema-new.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/l6/wy_sktzx7wb57l8grdq1j4zm0000gn/T/ipykernel_41611/195600819.py:1: DeprecationWarning: jsonschema.RefResolver is deprecated as of v4.18.0, in favor of the https://github.com/python-jsonschema/referencing library, which provides more compliant referencing behavior as well as more flexible APIs for customization. A future release will remove RefResolver. Please file a feature request (on referencing) if you are missing an API for the kind of customization you need.\n",
      "  from jsonschema import Draft202012Validator, RefResolver\n"
     ]
    }
   ],
   "source": [
    "from jsonschema import Draft202012Validator, RefResolver\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def validate_request(parameters):\n",
    "    # inject dummy values for missing fields - these will be added by the Dataportal during submission\n",
    "    parameters[\"datasetId\"] = \"Dataset Id\"\n",
    "    parameters[\"projectName\"] = \"Project Name\"\n",
    "    parameters[\"vcfLocations\"] = [\"vcfLocation1\", \"vcfLocation2\"]\n",
    "    # load validator\n",
    "    new_schema = \"./schemas/submit-dataset-schema-new.json\"\n",
    "    schema_dir = os.path.dirname(os.path.abspath(new_schema))\n",
    "    new_schema = json.load(open(new_schema))\n",
    "    resolveNew = RefResolver(base_uri=\"file://\" + schema_dir + \"/\", referrer=new_schema)\n",
    "    validator = Draft202012Validator(new_schema, resolver=resolveNew)\n",
    "    errors = []\n",
    "\n",
    "    for error in sorted(validator.iter_errors(parameters), key=lambda e: e.path):\n",
    "        error_message = f\"{error.message} \"\n",
    "        for part in list(error.path):\n",
    "            error_message += f\"/{part}\"\n",
    "        errors.append(error_message)\n",
    "    return errors\n",
    "\n",
    "validate_request(submission)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
